File: src/buildings/radar.js
import * as PIXI from 'pixi.js'
import { Building } from '../building.js'

export class Radar extends Building {

    constructor(game) {
        super('Radar', {x:4, y:4}, [[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]], game, false)
        this.sprite = PIXI.Sprite.from('Radar')

        this.radiusMax = 50
        this.currentRadius = 0
        this.growthChance = 0.1

        this.description = `Reveals tiles in an expanding radius. Has a 1 in 10 chance each tick to grow until radius ${this.getRadiusMax()}.`
    }

    getRadiusMax() {
        return this.radiusMax
    }

    getCurrentRadius() {
        return this.currentRadius
    }

    getCenterLocations() {
        if (!Number.isFinite(this.location.x) || !Number.isFinite(this.location.y)) {
            return []
        }

        const centerX = this.location.x + 1
        const centerY = this.location.y + 1
        return [
            { x: centerX, y: centerY },
            { x: centerX + 1, y: centerY },
            { x: centerX, y: centerY + 1 },
            { x: centerX + 1, y: centerY + 1 }
        ]
    }

    revealUnlockedArea(cave, previousRadius = -1) {
        if (!cave) {
            return 0
        }

        return cave.revealTilesBetweenRadii(this.getCenterLocations(), previousRadius, this.currentRadius)
    }

    onBuilt(cave) {
        cave.revealTiles(this.tileArray)
        this.revealUnlockedArea(cave, -1)
    }

    tick(cave) {
        if (!cave || this.currentRadius >= this.radiusMax) {
            return 0
        }

        if (Math.random() >= this.growthChance) {
            return 0
        }

        const previousRadius = this.currentRadius
        this.currentRadius++
        return this.revealUnlockedArea(cave, previousRadius)
    }
}

File: TriloGame.CSharp/src/TriloGame.Game/Core/Buildings/Radar.cs
using TriloGame.Game.Core.Simulation;
using TriloGame.Game.Shared.Math;
using TriloGame.Game.Shared.Utilities;

namespace TriloGame.Game.Core.Buildings;

public sealed class Radar : Building
{
    public Radar(GameSession session)
        : base("Radar", new GridPoint(4, 4), [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], session, false)
    {
        TextureKey = "Radar";
        Recipe = new Dictionary<string, int>(StringComparer.Ordinal) { ["Sandstone"] = 20 };
        RadiusMax = 50;
        CurrentRadius = 0;
        GrowthChance = 0.1;
        Description = $"Reveals tiles in an expanding radius. Has a 1 in 10 chance each tick to grow until radius {RadiusMax}.";
    }

    public int RadiusMax { get; }

    public int CurrentRadius { get; private set; }

    public double GrowthChance { get; }

    public IReadOnlyList<GridPoint> GetCenterLocations()
    {
        if (Location is null)
        {
            return [];
        }

        var centerX = Location.Value.X + 1;
        var centerY = Location.Value.Y + 1;
        return
        [
            new GridPoint(centerX, centerY),
            new GridPoint(centerX + 1, centerY),
            new GridPoint(centerX, centerY + 1),
            new GridPoint(centerX + 1, centerY + 1)
        ];
    }

    public int RevealUnlockedArea(World.Cave cave, int previousRadius = -1)
    {
        return cave.RevealTilesBetweenRadii(GetCenterLocations(), previousRadius, CurrentRadius);
    }

    public override void OnBuilt(World.Cave cave)
    {
        cave.RevealTiles(TileArray);
        RevealUnlockedArea(cave, -1);
    }

    public override int Tick(World.Cave cave)
    {
        if (CurrentRadius >= RadiusMax || RandomUtil.NextDouble() >= GrowthChance)
        {
            return 0;
        }

        var previous = CurrentRadius;
        CurrentRadius++;
        return RevealUnlockedArea(cave, previous);
    }
}
